From: Luke Deller <luke@deller.id.au>
Subject: [PATCH resend] server: implement file random access hint on OS X
Message-Id: <20220128110824.9831-1-luke@deller.id.au>
Date: Fri, 28 Jan 2022 22:08:24 +1100

Disable read ahead for files opened with FILE_RANDOM_ACCESS hint on
Mac OS X and FreeBSD.

Signed-off-by: Luke Deller <luke@deller.id.au>
---
 server/fd.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/server/fd.c b/server/fd.c
index 11111111111..11111111111 100644
--- a/server/fd.c
+++ b/server/fd.c
@@ -2289,6 +2289,13 @@ struct fd *open_fd( struct fd *root, const char *name, struct unicode_str nt_nam
         posix_fadvise( fd->unix_fd, 0, 0, POSIX_FADV_RANDOM );
         break;
     }
+#elif defined F_RDAHEAD /* Mac OS X, FreeBSD */
+    switch (options & (FILE_SEQUENTIAL_ONLY | FILE_RANDOM_ACCESS))
+    {
+    case FILE_RANDOM_ACCESS:
+        fcntl( fd->unix_fd, F_RDAHEAD, 0 );
+        break;
+    }
 #endif
 
     if (root_fd != -1) fchdir( server_dir_fd ); /* go back to the server dir */

-- 
2.39.0

